src/librustc/hir/lowering.rs

```diff
@@ -1149,13 +1149,79 @@ impl<'a> LoweringContext<'a> {
         }
     }
 
-    fn lower_generic_arg(&mut self,
-                         arg: &ast::GenericArg,
-                         itctx: ImplTraitContext<'_>)
-                         -> hir::GenericArg {
+    fn lower_generic_arg(
+        &mut self,
+        arg: &ast::GenericArg,
+        itctx: ImplTraitContext<'_>
+    ) -> hir::GenericArg {
         match arg {
             ast::GenericArg::Lifetime(lt) => GenericArg::Lifetime(self.lower_lifetime(&lt)),
-            ast::GenericArg::Type(ty) => GenericArg::Type(self.lower_ty_direct(&ty, itctx)),
+            ast::GenericArg::Type(ty) => {
+                // A path type argument could be an incorrectly parsed generic const argument. If
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
+                        if let Some(partial_res) = self.resolver.get_partial_res(ty.id) {
+                            match partial_res.base_res() {
+                                // FIXME(const-generics): Currently matches against all items in
+                                // the value namespace. Maybe this should be pared down?
```

Comment:
I wonder if we could automate getting the namespace so we don't have to match on the cases here.

Reply:
I looked at implementing a `is_valid_in_ns` function on `Res`, but didn't go with that implementation because I didn't want to leave out `Res::Err` or `Res::Def(DefKind::Ctor(_, CtorKind::Fictive)`. Both are valid in the value namespace (to my knowledge), but not here.

Thinking on it, I suppose I could have specifically checked for those two variants and still implemented the function... Don't code at 3 a.m., kids.